### PR TITLE
fix: S3 200 error customization should only not apply to streaming + blob shapes

### PR DESCRIPTION
**Pull request by dayaffe**
## Issue \#
#1629 

## Description of changes
- S3 200 Error customization should only not apply to streaming + blob shapes not all streaming shapes.
- This change only effects one method `selectObjectContent` which didn't have the customization applied but now does. The response of this method contains an event stream.
- We should only check the root of the response for an <Error> node not if <Error> is present anywhere in the response

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jbelkins** [comment]
so event streams DO get the s3/200 customization applied to them?

File: Package.swift
Comment (sichanyoo) [inline Package.swift:47]: nit: Convention for acronyms; use `XML` instead of `Xml`

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:25]: nit: `findNode` isn't used here, is it?

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:58]: Q: Am I understanding correctly: since the body can't be read twice, the read body is copied back in to the response for it to be available down the line in operation stack?

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:44]: nit: This isn't checking for error, it's just reading body

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:49]: I think comments in this highlighted block can just be deleted without harm since what's being done is pretty straightforward; unnecessary comments may actually confuse if it ever gets outdated in future.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt:41]: nit: Better name might be `outputIsNotAStreamingBlobShape`. As it stands, it reads to denote a shape that is `Not Streaming Shape And Is A Blob Shape` unless I read the next 4 lines of code

**Review by sichanyoo — CHANGES_REQUESTED** [review]
Re: comments and namings

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (dayaffe) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:58]: Yes otherwise it will be `nil` when the user tries to look at the events

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:26]: I like this, much more solid than to search a string for XML

Comment (jbelkins) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:26]: But, RestXML supports error wrapping.
https://smithy.io/2.0/aws/protocols/aws-restxml-protocol.html#aws-protocols-restxml-trait

Will this code detect the error whether or not it is wrapped?

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:48]: How do we know we will have enough data here to determine if the body returns an error?
`readData()` could return zero bytes if the data is not yet available, and the stream could be filled with error data after this step is complete.

**Review by jbelkins — COMMENTED** [review]
Questions inline

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (dayaffe) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:48]: we don't get the stream back all at once, each response is a separate action so it should be checking each streamed message for <Error> at root

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (dayaffe) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:26]: In short, no if it is not Error at the root it won't detect it.

**Review by sichanyoo — APPROVED** [review]
should be good to go pending approval from Josh also

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:26]: OK.  Reading the SEP, it sounds pretty clear that the affected errors will be in the root.

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:65]: Does this need to be public?  Is it used anywhere other than in the middleware above?

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:70]: Shouldn't the copied stream be created un-closed?  At this point we don't know if there will be more data in the stream in the future.

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (jbelkins) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:65]: Also, can this param be non-nil `Data`?  In the code above, `data` is safe-unwrapped before this function is called

Comment (dayaffe) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:65]: sure I can make it private and non-nil unless we want to make it generally available and move it to a place like SmithyStreams or ClientRuntime. Only issue with that is neither has both ByteStream and the various types of streams

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift
Comment (dayaffe) [reply Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Middlewares/AWSS3ErrorWith200StatusXMLMiddleware.swift:70]: If its created unclosed then the AsyncThrowingStream will never complete when looping through it. It yields each event until an end event and then hangs forever.

**Review by jbelkins — APPROVED** [review]

**Pull request merged**
